core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1252,11 +1536,221 @@ public UnifyResult apply(UnifyRuleCall call) {
+    for (RexNode rex: shuttle.apply(calc.program.getProjectList())) {
+      projects.add(rex);
+    }
+    if (condition == null) {
+      return Pair.of(calc.cluster.getRexBuilder().makeLiteral(true), projects);
+    } else {
+      return Pair.of(condition, projects);
+    }
+  }
+
+  /**
+   * Generate result by merging parent and child if they are both MutableCalc.
+   * Otherwise result is the child itself.
+   */
+  private static UnifyResult tryMergeParentCalcAndGenResult(
+      UnifyRuleCall call, MutableRel child) {
+    MutableRel parent = call.query.getParent();
+    if (child instanceof MutableCalc && parent instanceof MutableCalc) {
+      MutableCalc mergedCalc = mergeCalc(call.getCluster().getRexBuilder(),
+          (MutableCalc) parent, (MutableCalc) child);
+      if (mergedCalc != null) {
+        // Note that property of stopTrying in the result is false
+        // and this query node deserves further matching iterations.
+        return call.create(parent).result(mergedCalc, false);
+      }
+    }
+    return call.result(child);
+  }
+
+  /** Merge two MutableCalc together. */
+  private static MutableCalc mergeCalc(
+      RexBuilder rexBuilder, MutableCalc topCalc, MutableCalc bottomCalc) {
+    RexProgram topProgram = topCalc.program;
+    if (RexOver.containsOver(topProgram)) {
+      return null;
+    }
+
+    RexProgram mergedProgram =
+        RexProgramBuilder.mergePrograms(
+            topCalc.program,
+            bottomCalc.program,
+            rexBuilder);
+    assert mergedProgram.getOutputRowType()
+        == topProgram.getOutputRowType();
+    return MutableCalc.of(bottomCalc.getInput(), mergedProgram);
+  }
+
+  private static RexShuttle getExpandShuttle(RexProgram rexProgram) {
+    return new RexShuttle() {
+      @Override public RexNode visitLocalRef(RexLocalRef localRef) {
+        return rexProgram.expandLocalRef(localRef);
+      }
+    };
+  }
+
+  /** Check if condition cond0 implies cond1. */
+  private static boolean implies(
+      RelOptCluster cluster, RexNode cond0, RexNode cond1, RelDataType rowType) {
+    RexExecutorImpl rexImpl =
+        (RexExecutorImpl) (cluster.getPlanner().getExecutor());
+    RexImplicationChecker rexImplicationChecker =
+        new RexImplicationChecker(cluster.getRexBuilder(), rexImpl, rowType);
+    return rexImplicationChecker.implies(cond0, cond1);
+  }
+
+  /** Check if join condition only references RexInputRef. */
+  private static boolean referenceByMapping(
+      RexNode joinCondition, List<RexNode>... projectsOfInputs) {
+    List<RexNode> projects = new ArrayList<>();
+    for (List<RexNode> projectsOfInput: projectsOfInputs) {
+      projects.addAll(projectsOfInput);
+    }
+
+    try {
+      RexVisitor rexVisitor = new RexVisitorImpl<Void>(true) {
+        @Override public Void visitInputRef(RexInputRef inputRef) {
+          if (!(projects.get(inputRef.getIndex()) instanceof RexInputRef)) {
+            throw new Util.FoundOne(inputRef);
```

Comment:
I think it is a good candidate to use `Util.FoundOne.NULL`. What do you think, @julianhyde ?